pallets/gear/src/internal.rs

```diff
@@ -632,22 +673,53 @@ where
                     Self::remove_gas_reservation_with_task(dispatch.source(), reservation_id);
                 }
             }
-        }
 
-        if !dispatch.value().is_zero() {
-            // Reserving value from source for future transfer or unreserve.
-            CurrencyOf::<T>::reserve(&from, value)
-                .unwrap_or_else(|e| unreachable!("Unable to reserve requested value {:?}", e));
+            // `HoldBound` cost builder.
+            let hold_builder = HoldBound::<T>::by(CostsPerBlockOf::<T>::dispatch_stash());
+
+            // Calculating correct hold bound to lock gas
+            let maximal_hold = hold_builder.clone().maximum_for_message(dispatch.id());
+            let bn_delay = delay.saturated_into::<BlockNumberFor<T>>();
+            let hold = hold_builder.duration(bn_delay).min(maximal_hold);
+
+            // Locking funds for holding.
+            GasHandlerOf::<T>::lock(dispatch.id(), hold.lock())
+                .unwrap_or_else(|e| unreachable!("GasTree corrupted! {:?}", e));
+
+            if hold.expected_duration().is_zero() {
+                // TODO: Replace with unreachable call after:
+                // - `HoldBound` safety usage stabilized;
+                log::error!("Failed to figure out correct wait hold bound");
+                return;
+            }
+
+            if !dispatch.value().is_zero() {
+                // Reserving value from source for future transfer or unreserve.
+                CurrencyOf::<T>::reserve(&from, value)
+                    .unwrap_or_else(|e| unreachable!("Unable to reserve requested value {:?}", e));
+            }
         }
 
         // Saving id to allow moving dispatch further.
         let message_id = dispatch.id();
 
+        // Add block number of insertation
+        let start_bn = Self::block_number();
+        let finish_bn = start_bn.saturating_add(delay.into());
+
+        let delay_interval = Interval {
+            start: start_bn,
+            finish: finish_bn,
```

Comment:
```suggestion
            finish: hold_bound.expected(),
```

I'd even impl `fn to_interval(&self) -> Interval<BlockNumber>` for HoldBound

Reply:
done